### PR TITLE
Fix type and linkage errors with parallel codegen

### LINKED ISSUE #19826: Using `-C codegen_units=N` while bootstrapping rust generates an LLVM assertion
### STR

```
$ git clone https://github.com/rust-lang/rust
$ cd rust
$ mkdir build
$ cd build
$ ../configure
$ RUSTFLAGS=-Ccodegen_units=2 make
rustc: x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc
rustc: x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand
rustc: x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunicode
rustc: x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc
rustc: /home/rustbuild/src/rust-buildbot/slave/snap3-linux/build/src/llvm/lib/IR/Constants.cpp:1609: static llvm::Constant* llvm::ConstantExpr::getPointerCast(llvm::Constant*, llvm::Type*): Assertion `S->getType()->isPtrOrPtrVectorTy() && "Invalid cast"' failed.
Aborted (core dumped)
```

### Version

444fa1b

---

Sadly, I couldn't generate a minimal script that reproduces the problem. But these seem to be the required ingredients:

- Cross crate. The linked crate must use unboxed closures.
- There must be some `#[inline]`d function in the root crate.
- And, of course, the `-C codegen_units` flag.

cc @nikomatsakis @nick29581

### PR DESCRIPTION
**Pull request by nrc**
Closes #19826

r? @nikomatsakis 

There is still some work to do until parallel codegen is perfect - we don't pass make check due to missing the symbol for registering a macro. But we do bootstrap with codegen_units=4 and pass make check without codegen_units, so I think it should land.

For the curious, make rustc-stage2 -j8:

```
codegen_units=1:

real  9m18.074s
user  11m59.858s
sys 0m13.281s


codegen_units=4:

real  6m3.672s
user  13m5.474s
sys 0m12.146s
```

Which is a 33% speedup :-)

### CODE REVIEW

**Comment by bstrie** [comment]
Those speedups looks awesome! But I have to ask, has the single-threaded performance regressed compared to a compiler that doesn't have this patch?

**Comment by nrc** [comment]
@bstrie I haven't measured, but it really shouldn't have - nearly everything here is either specific to codegen_units or fixes actual bugs. If anything, single threaded performance should have slightly increased because we no longer generate some duplicate code.

**Comment by bstrie** [comment]
Awesome. I'd love to see someone try the same timing with Servo.

**Comment by nrc** [comment]
Me too! It has the most affect where crates are large and monolithic, i.e., like rustc used to be. We used to get 50% speedup on a stage2 build with || codegen. I believe Servo is many small crates, so the effect will probably be similar to rust or smaller.

**Comment by jdm** [comment]
The script crate definitely fits that description, and that's our biggest offender.

**Comment by nrc** [comment]
If anyone measures please let me know the times (and if it even works).

File: src/librustc_trans/trans/controlflow.rs [hunk 1]
```diff
@@ -171,16 +170,7 @@ pub fn trans_if<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
             // if true { .. } [else { .. }]
             bcx = trans_block(bcx, &*thn, dest);
             trans::debuginfo::clear_source_location(bcx.fcx);
-
-            if let Some(elexpr) = els {
-                let mut trans = TransItemVisitor { ccx: bcx.fcx.ccx };
-                trans.visit_expr(&*elexpr);
-            }
```
Comment (alexcrichton) [inline src/librustc_trans/trans/controlflow.rs]: I vaguely remember that a long time ago this was required for something like:

```rust
// foo.rs
pub fn foo<T>() {
    if true { 
    } else {
        println!("test");
    }
}

// bar.rs
extern crate foo;
fn main() {
    foo::foo<int>();
}
```

The monomorphization of `foo` in `bar.rs` caused something to reference the statics produced by `println!`, assuming they were available in `foo.o` (but we didn't translate them due to the constant branch).

I may also be remembering something totally bogus though! I suspect if this passes all tests it's fine anyway.

Comment (nrc) [inline src/librustc_trans/trans/controlflow.rs]: I believe we should still translate the dead branch thanks to the visit from base::trans_item, so the extra visit here is unnecessary

Comment (alexcrichton) [inline src/librustc_trans/trans/controlflow.rs]: Aha, right!

**Comment by nikomatsakis** [comment]
@nrc r+ modulo possible error message changes. Nice work!

**Comment by nrc** [comment]
That looks like a real error, but I cannot reproduce locally

**Comment by alexcrichton** [comment]
Ah I've seen that brand of failure before, so I think it's safe to say it's benign (ish).

**Pull request merged**